Add category backups management functionality

1katoda opened the pull request:
Add functionality to this plugin, which allows backups management on a category context to all users with the appropriate capabilities, which are by default available to all the roles with the archetype of **manager**. The users are only able to manage backups on courses, on which they have the required capability.
This commit also adds an additional plugin settings page with a database query optimization option, which greatly reduces the amount of time it takes for the query to execute, but decreases accuracy of finding the backup files to only those backups, which have been created on the current Moodle instance (no user or plugin uploaded files). The increased query execution speed helps on large Moodle installations with a lot of files.
Also added slovenian translations and a changelog with a more descriptive list of changes.

danmarsden commented:
thanks @1katoda - I'll try to review this soon. If you get a chance, can you please remove the 
Fix Moodle 3.11 get_all_user_name_fields deprecation patch from your PR? - personally I prefer to create new branches rather than introduce version checks like that so I'll look at creating a new 3.10 and lower branch and then have the 3.11 and higher branches supported in the main branch. - we'll then just drop in the new api change to the main branch without needing to support older branches.

thanks!

danmarsden commented:
Hi @1katoda - Nice work - looks like a useful feature!

I'm not a big fan of the amount of duplication required in the new categorybackups.php file - it looks like a lot of that stuff should be in a class or some local functions, but I'm also not sure we actually need a new file here? - I'd be more inclined to make the core index.php file allow use within course categories.

No worries if you don't have time to implement that- it just prevents us from being able to include it in the main branch here at the moment.

Also - can you please remove the extra lang file you've added here? - the right place for language translation for plugins that are published in Moodle's plugins db is in Moodle's AMOS tool - not in the git repo itself.

thanks heaps for working on this and sharing it here!

danmarsden commented:
It would also be helpful if you separated out the features you've got here into different Pull requests - I see some stuff around filtering the backups based on "site-level" settings - I'm inclined to think these might be more useful as preferences on the report itself rather than site level settings, and it would be a lot easier to review this as a separate pull request from your "category backups" feature.
thanks!

1katoda commented:
Hello,
thanks a lot for your review and your feedback.
I cannot, however, take all the credit for this work, as it was a joint effort between me and my colleague at ARNES (slovenian NREN).

Of course, I will remove the `Fix Moodle 3.11 get_all_user_name_fields deprecation` patch, I added it because the failing pipeline made me nervous.

I will also remove the translations from the PR (sorry; I did not know that the plugin translations are also handled by AMOS, thank you for the information), as well as the duplicated functionality in `categorybackups.php`. May I/we move all of the output functionality into a separate class, to be applied by a plugin renderer? I think the code and the functionalities might be easier to read and review that way.

The additional site-level settings affecting the backups search query stem from our large databases inability to execute it quickly enough. I think the *"Moodle backups only"* option should be left in the site settings section for the site administrators, since the end-users with the correct permissions on a category context won't benefit from the additional option in the report, if the query fails to execute in time and the page reaches timeout. Perhaps I could add a global site setting, which would enable the ability for the users to choose their query, but still leave it in the hands of administrators to disable the option alltogether?
As for the *"Include activities"* option - I agree, that one should probably be left in the hands of users. Both will be modified in a separate pull request and removed from this one.

Thank you again for the review, please let me know, if you have any pointers or objections to what I have written.

Best regards

1katoda commented:
I have created separate pull requests for both the functionalities, so this PR is no longer relevant, except for the review aspect.  
Closing.
